src/graph.rs

```diff
@@ -3426,19 +3426,38 @@ impl<'a, 'graph> Builder<'a, 'graph> {
     package_req: &PackageReq,
   ) -> Option<Version> {
     // try to find in the list of existing versions first
-    if let Some(existing_versions) =
+    let versions = if let Some(existing_versions) =
       self.graph.packages.versions_by_name(&package_req.name)
     {
-      if let Some(version) = resolve_version(
-        &package_req.version_req,
-        existing_versions.iter().map(|nv| &nv.version),
-      ) {
-        return Some(version.clone());
-      }
+      existing_versions
+        .iter()
+        .map(|nv| &nv.version)
+        .collect::<Vec<_>>()
```

Comment:
I think we should keep the existing code here and prioritize existing versions even if they're yanked because it looks like this would cause the lockfile to not be respected for yanked packages (ex. someone has a package locked, someone yanks a package and publishes a malicious package version, now you're compromised) and also we should prefer to not cause duplicate dependencies by preferring yanked packages that have previously been resolved. The user will get a warning and then they can decide how to proceed.

Reply:
But in the lockfile case the code after this will be a filtration of this, which should be a 1-element vec. So if it's yanked, `unyanked_versions` will be empty, right?

Reply:
Ah, I just misread the code